### DISCUSSION TITLE
Mapping protein FastA to gene information

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1228] [Q&A] [answered]
Starting a new discussion (follow ups from [this one about `blastx`](https://github.com/RobertsLab/resources/discussions/1226) and [this one about Uniprot ID mapping](https://github.com/RobertsLab/resources/discussions/1227)).

## WHAT I'VE TRIED

Initially, I thought I could get Uniprot Accession information by using `DIAMOND blastx` to annotate the genome. This did not work for a variety of reasons (see [this discussion](https://github.com/RobertsLab/resources/discussions/1226)). I also tried running `DIAMOND blastx` without the index or block size information, and got some recommendations (`-b12 -c1`). I'm running it again with those recommendations now.

I then tried using the Uniprot ID mapping script from @kubu4 (see [this discussion](https://github.com/RobertsLab/resources/discussions/1227)). We realized this wasn't working because the Uniprot database does not include the information from the new genome annotation! I will [work from the protein annotation FastA for ID mapping](https://github.com/RobertsLab/resources/discussions/1227#discussioncomment-817230) and will report back here.

## MY NEW PROBLEM

Assuming the ID mapping from the protein FastA works, how do I map the protein IDs to gene IDs? Looking at the [annotation README](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/902/806/645/GCF_902806645.1_cgigas_uk_roslin_v1/README.txt), I don't see any RefSeq files that may help with this. Any suggestions?

**Comment by kubu4** [comment]
Looks like RefSeq accession can be reliably mapped to NCBI gene IDs:

e.g. `NM_001305288` only maps to `105326593`

This info is in the GFF. Using GFFutils,  you'll need to operate on `mRNA` instead of `gene`. Then, you'll have to strip off the beginning notation of the ID results.

e.g.

```bash
gtf_extract \
--feature mRNA \
--fields=chrom,start,end,ID \
--gff ${data_dir}/${orig_gff} \
| sed 's/rna-//'
```
Then, you can proceed with isolating column 4 to generate your list. Then, map using the updated Perl script to use the RefSeq protein accession instead of the NCBI gene ID.

EDITED: Remove pound symbol in front of column number to remove autolink to GitHub Issue.

**Reply by yaaminiv** [reply]
To clarify, I these are the three changes I would make from your original [salmon genome script](https://nbviewer.jupyter.org/github/RobertsLab/code/blob/master/notebooks/sam/20210601_ssal_gff-annotations.ipynb):

- use the genome.gff
- modify gft_extract to run on mrNA
- map with the updated perl script

**Reply by kubu4** [reply]
Yes. You might need to have the Perl script use REFSEQ_NT_ID instead of P_REFSEQ_AC to work on nucleotides instead of proteins.

**Reply by yaaminiv** [reply]
Okay! I have some new errors. 

Here is my gene list: https://github.com/RobertsLab/project-oyster-oa/blob/master/analyses/Haws_08-GOterm-annotation/20210602_cgigas_roslin_gene-list.txt

I ran my perl script with `REFSEQ_NT_ID`, no output. I got the following error for each line of my gene list:

> Semicolon seems to be missing at 20210602_cgigas_roslin_gene-list.txt line 1

Not sure what semicolon I need to have for input!

*deleted previous comment because it was on the wrong thread*

**Reply by kubu4** [reply]
> I ran my perl script with REFSEQ_NT_ID

Have you tried with `P_REFSEQ_AC`?

Per the semicolon error message, maybe it doesn't like accessions ending with `.1`?

Add this to end of your `gtf_extract` command to remove that:

`| sed 's/.1$//`

**Reply by yaaminiv** [reply]
> Have you tried with P_REFSEQ_AC

Tried and got the same error!

> Per the semicolon error message, maybe it doesn't like accessions ending with .1?

Trying without that now.

**Comment by yaaminiv** [comment]
> I will work from the protein annotation FastA for ID mapping and will report back here.

Extracted protein IDs from the protein annotation FastA, got 0 results:

```
Failed, got 500 Server closed connection without sending any data back for https://www.uniprot.org/uploadlists/
```

Will try working from the GFF

**Reply by kubu4** [reply]
You might need to have the Perl script use `REFSEQ_NT_ID` instead of `P_REFSEQ_AC` to work on nucleotides instead of proteins.

**Comment by sr320** [comment]
FWIW, 
I would grab RefSeq IDs from Fasta.
Upload to web, download associated info.
It might be the case that your RefSeq ID does not come back down with annotation (eg GO etc)
In that case I would then take Salmon Fasta and blastp (not diamond) to fasta of corresponding mapped UniProtKB seqs.

**Reply by yaaminiv** [reply]
@sr320 To clarify: 

- grab RefSeq IDs from protein Fasta and use web interface
- `blastp` Roslin genome fasta to something else...?

**Reply by sr320** [reply]
> grab RefSeq IDs from protein Fasta and use web interface 

yes

> blastp Roslin genome fasta to something else...?

that is protein fasta only if 

>It might be the case that your RefSeq ID does not come back down with annotation (eg GO etc)

Only if this is case.. then you would need to know which RefSeq IDs correspond with annotations, and I would use blastp and the UniProtKB sequences found in mapping as db.

happy to walk through in zoom if helpful

**Comment by yaaminiv** [comment, marked as answer]
Uniprot GUI mapping yielded data for ~1000 entries, only UniParc mapping https://d.pr/i/jXB0Y9

Decided to just use `blastx` to annotate the genome in order to get GOterms for enrichment
